[a11y] Don't wrap legend tags in divs

bkjohnson opened the pull request:
## Description

This _might_ be a total fix for https://github.com/department-of-veterans-affairs/va.gov-team/issues/767

There is some refactoring happening but I scaled back on that to keep the focus mostly on the accessibility issue. The wrapping divs around the `<legend>` are removed, but further testing still needs to be done to make sure that the styling is acceptable.


## Testing done

Local

## Screenshots


## Acceptance criteria
- [ ] `<legend>` tags are not wrapped inside of a `<div>`

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

bkjohnson committed:
Remove optimization

I'm trying to refactor this into a functional component, so I don't need
this right now

bkjohnson committed:
Simplify isRequired

schemaRequired is already a boolean, and if it is true we will return
it, otherwise we will return false.  We can slim it down

bkjohnson committed:
Remove unused onPropertyBlur method

bkjohnson committed:
Remove state & getStateFromProps

State appears unused here

bkjohnson committed:
Move orderAndFilterProperties out of the class

It doesn't need to be a method due to how the arguments are set up

bkjohnson committed:
Convert to functional component

This is a bit risky...

bkjohnson committed:
Revert "Convert to functional component"

This reverts commit 03b618b2a2d343cbddda628380ebca368df2811d.

bkjohnson committed:
Remove wrapping divs

Still need to consider the styling

bkjohnson commented on src/platform/forms-system/src/js/fields/ObjectField.jsx:
```diff
@@ -107,50 +118,11 @@ class ObjectField extends React.Component {
     };
   }
 
-  onPropertyBlur(name) {
-    return (path = []) => {
-      this.props.onBlur([name].concat(path));
-    };
-  }
```
This is never called - I removed the bind for it as well.

bkjohnson commented on src/platform/forms-system/src/js/fields/ObjectField.jsx:
```diff
@@ -107,50 +118,11 @@ class ObjectField extends React.Component {
-    }
-
-    return false;
+    return (
+      Array.isArray(schema.required) && schema.required.indexOf(name) !== -1
+    );
```
This is just simplified from the following, where `isTrue` represents the returned value on line 124:

```js

const isTrue = true;

if (isTrue) {
  return isTrue;
}

return false
```

bkjohnson committed:
Merge branch 'master' into 767-axe-refactor

bkjohnson commented on src/platform/forms-system/src/js/fields/ObjectField.jsx:
```diff
@@ -51,14 +74,12 @@ class ObjectField extends React.Component {
 
   constructor(props) {
     super(props);
-    this.state = this.getStateFromProps(props);
```
state isn't used in this component, so this function can be removed as well.

bkjohnson commented on src/platform/forms-system/src/js/fields/ObjectField.jsx:
```diff
@@ -226,9 +198,9 @@ class ObjectField extends React.Component {
 
     const fieldContent = (
-      <div className={containerClassNames}>
+      <>
         {hasTitleOrDescription && (
-          <div className="schemaform-block-header">
+          <>
```
Removing these `<div>`s should fix the accessibility error by preventing a `<div>` from being the top level component inside of a `<fieldset>`, which would happen when the schema had a title.

bkjohnson committed:
Merge branch 'master' into 767-axe-refactor

bkjohnson committed:
Add the optimization back in

bkjohnson committed:
Add onPropertyBlur back

A unit test relies on calling this method to pass, and I don't want to
dig too much into that for this PR

bkjohnson committed:
Put a11y change behind an environment flag

bkjohnson commented on src/platform/forms-system/src/js/fields/ObjectField.jsx:
```diff
@@ -290,11 +283,84 @@ class ObjectField extends React.Component {
       </div>
     );
 
+    const accessibleFieldContent = (
+      <>
+        {hasTitleOrDescription && (
+          <>
```
This whole thing is a copy of `fieldContent`, but the only difference is these two fragments being here instead of being `<div>`s with className attributes.

bkjohnson replied:
The `fieldContent` I'm talking about:

https://github.com/department-of-veterans-affairs/vets-website/blob/e05f44309e62b09996f95977befd841e0c667220/src/platform/forms-system/src/js/fields/ObjectField.jsx#L255-L315

bkjohnson commented on src/platform/forms-system/src/js/fields/ObjectField.jsx:
```diff
@@ -290,11 +283,84 @@ class ObjectField extends React.Component {
-    return <div className={fieldsetClassNames}>{fieldContent}</div>;
+    return (
+      <div className={fieldsetClassNames}>
+        {environment.isProduction() ? fieldContent : accessibleFieldContent}
```
Environment flag here

bkjohnson commented on src/platform/forms-system/src/js/fields/ObjectField.jsx:
```diff
@@ -290,11 +283,84 @@ class ObjectField extends React.Component {
-      return <fieldset className={fieldsetClassNames}>{fieldContent}</fieldset>;
+      return (
+        <fieldset className={fieldsetClassNames}>
+          {environment.isProduction() ? fieldContent : accessibleFieldContent}
```
Environment flag here

bkjohnson committed:
Delete commented bit

1Copenut commented:
@bkjohnson I took a look at your feature branch last night, and had one question before I add my PR. When I ran through the education benefits form, I noticed a few instances where we had fieldset + legend patterns embedded inside another fieldset. While this is not a defect (axe validates), it's not a best practice according to the GOV.UK article [Using the fieldset and legend elements](https://accessibility.blog.gov.uk/2016/07/22/using-the-fieldset-and-legend-elements/)

It looks like this is an item outside the scope of your work here, so I'll review the PR for my small part.

Please advise if I should open a follow on ticket for exploring refactoring to get the nested fieldset refactored.

---
![Screen Shot 2020-08-18 at 5 17 21 PM](https://user-images.githubusercontent.com/934879/90658511-477a4580-e209-11ea-8432-2a641ff08465.png)

bkjohnson commented:
Opening a separate ticket for that sounds like a good idea. I'll take a look and see if I can understand it a bit more though.

1Copenut commented:
Please see my comment on issue 12718 about how we could unwind the nested fieldset + legend: https://github.com/department-of-veterans-affairs/va.gov-team/issues/12718#issuecomment-679377527

I hate to stop progress on this PR any longer, but I feel the nesting could cause a lot of additional problems for JAWS and NVDA if we remove the extra divs without a plan to refactor the parent markup.

/cc @emilywaggoner && @meganhkelley for awareness

rianfowler commented on src/platform/forms-system/src/js/fields/ObjectField.jsx:
```diff
@@ -38,6 +38,29 @@ function setFirstFields(id) {
+// This runs a series of steps that order properties and then group them into
+// expandable groups. If there are no expandable groups, then the end result of this
+// will be an array of single item arrays
+function orderAndFilterProperties(schema, uiSchema) {
```
you could also make it a static method

timwright12 commented:
@bkjohnson Looks like there's a conflict here and CI failing. Do we we still need this? If not can we close it out?

bkjohnson commented:
As far as I know the problem still exists, but it hasn't been a priority.

timwright12 commented:
@bkjohnson still need this PR?

bkjohnson committed:
Merge branch 'master' into 767-axe-refactor

bkjohnson committed:
Clean up merge

bkjohnson committed:
Undo out of scope refactoring

I don't like leaving these things here, but I want to limit the PR to
just the a11y issue

bkjohnson committed:
Add unit test

bkjohnson reviewed (commented):
This is mostly copy-paste with some minor changes. Once this has been tested & verified in staging we will need a followup PR to make this change take effect in production.
